deployment/terraform/keycloak.go

```diff
@@ -0,0 +1,279 @@
+package terraform
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/mattermost/mattermost-load-test-ng/deployment"
+	"github.com/mattermost/mattermost-load-test-ng/deployment/terraform/assets"
+	"github.com/mattermost/mattermost-load-test-ng/deployment/terraform/ssh"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+)
+
+func (t *Terraform) setupKeycloak(extAgent *ssh.ExtAgent) error {
+	keycloakBinPath := "/opt/keycloak/keycloak-" + t.config.ExternalAuthProviderSettings.KeycloakVersion + "/bin"
+
+	mlog.Info("Configuring keycloak", mlog.String("host", t.output.KeycloakServer.PublicIP))
+
+	command := "start-dev"
+	extraArguments := []string{}
+
+	if !t.config.ExternalAuthProviderSettings.DevelopmentMode {
+		command = "start"
+	}
+
+	sshc, err := extAgent.NewClient(t.output.KeycloakServer.PublicIP)
+	if err != nil {
+		return err
+	}
+
+	// Check if the keycloak database exists
+	result, _ := sshc.RunCommand(`sudo -iu postgres psql -l | grep keycloak 2> /dev/null`)
+	if len(result) == 0 {
+		mlog.Info("keycloak database not found, creating it and associated role")
+		_, err = sshc.RunCommand(`sudo -iu postgres psql <<EOSQL
+		CREATE USER keycloak WITH PASSWORD 'mmpass';
+		CREATE DATABASE keycloak OWNER keycloak;
+		GRANT ALL PRIVILEGES ON DATABASE keycloak TO keycloak;
+		EOSQL
+		'`)
+		if err != nil {
+			return fmt.Errorf("failed to setup keycloak database: %w", err)
+		}
+
+		if t.config.ExternalAuthProviderSettings.KeycloakDBDumpURI != "" && t.config.ExternalAuthProviderSettings.KeycloakRealmFilePath == "" {
+			err := t.IngestKeycloakDump()
+			if err != nil {
+				return fmt.Errorf("failed to ingest keycloak dump: %w", err)
+			}
+		}
+	}
+
+	// Check if we should use a custom dump, a custom realm file or the default one
+	if t.config.ExternalAuthProviderSettings.KeycloakRealmFilePath != "" {
+		_, err := sshc.UploadFile(t.config.ExternalAuthProviderSettings.KeycloakRealmFilePath, "/opt/keycloak/keycloak-"+t.config.ExternalAuthProviderSettings.KeycloakVersion+"/data/import/mattermost-realm.json", true)
```

Comment:
The destination realm file is copied in two places. Let's consolidate into a single variable.